pyoptsparse/pySNOPT/pySNOPT.py

```diff
@@ -328,14 +331,29 @@ def __call__(
                     raise Error("Failed to properly open %s, ierror = %3d" % (SummFile, ierror))
 
             # Calculate the length of the work arrays
-            # --------------------------------------
+            # ---------------------------------------
             nvar = self.optProb.ndvs
-            lencw = 500
-            leniw = 500 + 100 * (ncon + nvar)
-            lenrw = 500 + 200 * (ncon + nvar)
+            lencw = self.getOption("Total character workspace")
+            leniw = self.getOption("Total integer workspace")
+            lenrw = self.getOption("Total real workspace")
 
-            self.setOption("Total integer workspace", leniw)
-            self.setOption("Total real workspace", lenrw)
+            # Set a flag to avoid overwriting user-specified lengths
+            if lencw is None and leniw is None and lenrw is None:
+                defaultLengths = True
```

Comment:
I would prefer a more descriptive variable name such as `checkLengths`, since it's possible for some of these to use default values while others are set by the user.